### PR TITLE
Clean up console code

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
These are features we haven't been supporting already (for PowerShell Core) and no plans to support as they exist for legacy reasons only for Windows PowerShell:

- removed sta/mta switches and code
- removed psconsolefile switch and code
- removed importsystemmodules switch and code
- removed font changing code
- merged `#if CORECLR` code

Fix https://github.com/PowerShell/PowerShell/issues/4987
Fix https://github.com/PowerShell/PowerShell/issues/4350

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleShell.cs [hunk 1]
```diff
@@ -25,18 +24,6 @@ public static int Start(string bannerText, string helpText, string[] args)
         {
             return Start(null, bannerText, helpText, null, args);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleShell.cs]: Do we use first parameter? Can we remove it?

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleShell.cs]: The first param is RunspaceConfiguration.  It's being removed as part of https://github.com/PowerShell/PowerShell/pull/4942

Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleShell.cs]: Thanks! 

Closed.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 2]
```diff
@@ -2937,23 +2893,7 @@ private class ConsoleHostStartupException : Exception
...
 
-#if CORECLR
         internal Lazy<TextReader> ConsoleIn { get; } = new Lazy<TextReader>(() => Console.In);
-#else
-        internal Lazy<TextReader> ConsoleIn { get; } = new Lazy<TextReader>(() =>
-        {
-            // This is a workaround for a full clr issue that causes a hang when calling PowerShell from ruby.
-            // They use named pipes instead of anonymous pipes, and for some reason that triggers a hang
-            // reading from Console.In.
-            var inputHandle = ConsoleControl.NativeMethods.GetStdHandle(-10);
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: It seems like we should also remove NativeMethods.GetStdHandle() declaration since it is only used here.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 3]
```diff
@@ -145,11 +145,6 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
 
         private IntPtr GetMainWindowHandle()
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: Since we now always prompt using console we can get rid of this method since it will never be used.

Comment (PaulHigin) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: PromptForCredential() can be simplified since it now always prompts from console.  Also we can remove the HostUtilities.CredUIPromptForCredential() method since it also will no longer be used.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs [hunk 4]
```diff
@@ -163,42 +158,13 @@ private IntPtr GetMainWindowHandle()
             }
 
             return mainWindowHandle;
-#endif
         }
 
         // Determines whether we should prompt using the Console prompting
         // APIs
         private bool PromptUsingConsole()
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: We can remove this method and simply the conditional code in the two places it is called.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 5]
```diff
@@ -2830,20 +2727,7 @@ internal static void SetConsoleTextAttribute(ConsoleHandle consoleHandle, WORD a
...
             const uint OEM_CHARSET = 255;
-            // Suppress the PreFAST warning about not using Marshal.GetLastWin32Error() to
-            // get the error code.
-#pragma warning disable 56523
-#if CORECLR // TranslateCharsetInfo exists in an extension API set 'ext-ms-win-gdi-font-l1-1-1.dll', which is not available in NanoServer.
             return OEM_CHARSET;
-#else
-            CHARSETINFO csi;
-            const DWORD TCI_SRCCODEPAGE = 2;
-            if (!NativeMethods.TranslateCharsetInfo((IntPtr)codePage, out csi, TCI_SRCCODEPAGE))
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: We should remove all unused NativeMethods declarations (TranslateCharsetInfo, GetConsoleWindow, etc.).

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 6]
```diff
@@ -2830,20 +2727,7 @@ internal static void SetConsoleTextAttribute(ConsoleHandle consoleHandle, WORD a
         private static uint CodePageToCharSet(uint codePage)
         {
             const uint OEM_CHARSET = 255;
```
[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: Makes sense, will change

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 2 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Will remove

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs (hunk 3 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: Agree, will remove

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs (hunk 4 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs]: Will change

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs (hunk 5 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: Will remove

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by iSazonov** [comment]
@SteveL-MSFT Is the PR ready to merge?

**Comment by SteveL-MSFT** [comment]
@iSazonov since this is a breaking change (from Windows PowerShell), let me get an ack from @PowerShell/powershell-committee

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 7]
```diff
@@ -1893,22 +1893,7 @@ private static bool ReadConsoleOutputCJKSmall
                     }
                     else
                     {
-                        int charLength = LengthInBufferCells(contents[r, c].Character);
-                        if (charLength == 2)
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: As I understand this code - somebody is screen scraping, there is a wide character (2 cells), and the font in use can't render the character, so conhost used 1 cell.
This seems like an extreme corner case, but the code seems correct to me and I don't any value in removing the code.
Also - the PR description should be improved - there is no link to an issue providing more detail.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 8]
```diff
@@ -2944,103 +2783,9 @@ private static int LengthInBufferCellsFE(char c, ref HWND hwnd, ref HDC hDC, ref
...
-        /// <summary>
-        /// From IsConsoleFullWidth in \windows\core\ntcon\server\dbcs.c
-        /// </summary>
-        /// <param name="c"></param>
-        /// <param name="codePage"></param>
-        /// <returns></returns>
-        [SuppressMessage("Microsoft.Usage", "CA1806:DoNotIgnoreMethodResults",
-            MessageId = "Microsoft.PowerShell.ConsoleControl+NativeMethods.ReleaseDC(System.IntPtr,System.IntPtr)")]
-        private static int LengthInBufferCells(char c, uint codePage)
-        {
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: I think we still need LengthInBufferCells - but if you want to make a simplifying assumption when you can't know the font, you can use this [code](https://github.com/lzybkr/PSReadLine/blob/8edcf64663d7b7de3d9ef23435473e6529954b1e/PSReadLine/Render.cs#L512) from PSReadline.

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: I'll just use the code from PSReadline.  I ended up removing all this code as the fundamental assumption in the code is that we are using a specific codePage so all the calculations ended up returning a width of 1 anyways as all the other supporting code only applied to Windows PowerShell.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs (hunk 7 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: I'll update the PR description as I did not originally intend for this PR to be so extensive.  I simply started down the path of removing the -mta/-sta code since we weren't using it and then noticed the ConsoleFile code which wasn't being used, then the Font code... ugh...

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 9]
```diff
@@ -1682,32 +1657,19 @@ private void DoCreateRunspace(string initialCommand, bool skipProfiles, bool sta
...
-            // On NanoPS: always 'false' = default('false'=MTA) + NO possibility of overload through cmdline parameter '-mta'
-            // ThreadOptions should match on FullPS and NanoPS for corresponding ApartmentStates.
-            if (staMode)
-            {
-                // we can't change ApartmentStates on CoreCLR
-#if !CORECLR
-                runspace.ApartmentState = ApartmentState.STA;
-#endif
-                runspace.ThreadOptions = PSThreadOptions.ReuseThread;
-            }
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Is this removal intentional? I think if that's not set, we create a new thread every time you execute a new command line.

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Will put back the ReuseThread line

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 2 above)
Comment (lzybkr) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: You should verify the hang doesn't repro in Core. It looks like I did not add a test (probably because the simplest repro required installing Ruby on Windows), but with all the changes in the CLR, it seems prudent to verify we still don't need this code before removing.

---
In reply to: [143062595](https://github.com/PowerShell/PowerShell/pull/4995#discussion_r143062595) [](ancestors = 143062595)

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs [hunk 10]
```diff
@@ -1405,7 +1405,7 @@ int LengthInBufferCells(string s, int offset)
         public override
         int LengthInBufferCells(char c)
         {
-            return ConsoleControl.LengthInBufferCells(c);
+            return 1;
         }
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs]: This is wrong, plain and simple.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs [hunk 11]
```diff
@@ -322,24 +322,12 @@ public override
                 // mind is lost. This can lead to a malicious
                 // server prompting the user for a request
                 // which can appear to come from locally.
-                if (!PromptUsingConsole() && desc.ModifiedByRemotingProtocol)
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs]: Deleting this code makes the comment above very confusing.
It also implies you are introducing a potential security risk.

**Review by lzybkr — CHANGES_REQUESTED** [review]
There are several distinct changes in this PR, some of which don't feel like "clean up".

I think the commits should be cleaned up with better messages (reset and `git add -p`) and when merging, we should not squash this PR.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 9 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Actually, this line was only within the `staMode` which was not enabled for the user to set.  So currently, we use the default which is to use a new thread for each pipeline.  Do we want to change this behavior?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 2 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: This code wasn't being used since it's in the `#else` statement.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs (hunk 10 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs]: WIll fix!

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs (hunk 11 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfacePrompt.cs]: Will remove the comment since it only applied to the GUI prompt.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 9 above)
Comment (lzybkr) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: `staMode = true` was the default even if the user couldn't set it, right?
At any rate, I do think we want to reuse the thread.

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: No, we defaulted to MTA in PSCore

Comment (lzybkr) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: I see. We defaulted to STA on Windows for many years, so reusing the thread seems preferable.

**Comment by SteveL-MSFT** [comment]
@lzybkr cleaned up the commits to be distinct groups of work

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 12]
```diff
@@ -1515,10 +1514,6 @@ private uint Run(string bannerText, string helpText, bool isPrestartWarned, stri
 
                 _runspaceRef.Runspace.Close();
                 _runspaceRef = null;
-                if (staMode) // don't recycle the Runspace in STA mode
-                {
-                    ShouldEndSession = true;
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: If I understand correctly - we default to STA and remove the option to use MTA, correct?

Assuming so, this line should probably stay, but I don't quite understand why it's here.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 9 above)
Comment (lzybkr) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: @SteveL-MSFT - are you putting it back?

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 12 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: No, we defaulted to MTA and removing option to use STA

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs [hunk 13]
```diff
@@ -3321,6 +3050,39 @@ internal static void SetConsoleCursorInfo(ConsoleHandle consoleHandle, CONSOLE_C
 
 #endregion helper
 
+#region 
+
+        private static int LengthInBufferCells(char c)
+        {
+            if (c < 256)
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: This condition only applies to PSReadline.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 9 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: I need to verify something on Windows (was using a Mac for the changes) before I put it back.  With my other uber PR to remove apartmentstate all-up, the WSMan provider didn't work correctly.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 2 above)
Comment (lzybkr) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: @SteveL-MSFT - I know. That doesn't change my feedback. The code was originally added for Full CLR, but with the significant work that went into Core CLR 2.0, I think we need to revisit the issue in case the code is needed in Core now.

If you remove this code, you'll lose the context, and if it's still an issue, investigation will be much more difficult because nobody will remember the change I made that introduced the problem in the first place.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs (hunk 9 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: Looks good on my Windows box, will run through the tests and push the change.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs (hunk 13 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs]: will remove

**Comment by SteveL-MSFT** [comment]
@lzybkr I believe your feedback has been addressed, can you take another look?  Thanks

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee discussed this and was fine with all the changes except for removing the STA code.  Resolution is to leave the STA code commented out in case we have valid use cases on Windows on the future requiring support.

**Comment by SteveL-MSFT** [comment]
I believe all feedback has been addressed.  @lzybkr can you take another look when you get a chance?  Thanks.

**Comment by lzybkr** [comment]
>         {

This code is not called anymore, and is a slight variant of the version you added from PSReadLine below (which my version should be fine, it's more recent and has better comments.) :)


---
Refers to: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs:2736 in 6ca5d51. [](commit_id = 6ca5d517829587785df88ca7c9326adef577c11d, deletion_comment = False)

**Comment by SteveL-MSFT** [comment]
@lzybkr removed that method

**Review by lzybkr — APPROVED** [review]

**Comment by iSazonov** [comment]
@SteveL-MSFT Is the PR ready to merge?

**Comment by SteveL-MSFT** [comment]
@iSazonov yes, please!

**Comment by iSazonov** [comment]
It seems here we should preserve commits by create a merge commit?

**Comment by SteveL-MSFT** [comment]
@iSazonov yes, I believe that is the case
